Sql

cheungkinwong opened the pull request:
I'm using php admin to create my database structure.
create a database called 'database_note' if it doesn't already exist before.
Then create the table 'notes' within 'database_note' if it doesn't exist.
there are 4 columns in 'notes': 
title which is the primary and has a limit of 40 character.
author also has a 40 character limit.
notes which will contain the message, this does not have a limit.
dates which will add a timestamp, it can't be empty but it will add the timestamp by default and will update the timestamp on update, so we can see when the last edit of the note is.

I called my database 'database_note' to specify and avoid conflict in naming.
title and author has a limit of 40 to keep the title from being too long.
I gave title the PRIMARY KEY constraint to uniquely identifies each record in a table.

cheungkinwong committed:
add sql code to create database structure

cheungkinwong committed:
added primary key

cheungkinwong committed:
removed stuff i didn't understand

SeppeVerhavert reviewed (commented):
Looks good. What happens if you have 2 titles that are the same?

YelenaMerzlyakova reviewed (commented):
It looks like a good query for a SQL Database. I was given a good explanation for the choices presented

Lisonallie commented:
Looks good. Could you explain why you chose things like the names you chose, the lengths of fields, and why 'Titles' and 'Author' are 'NOT NULL'?

cheungkinwong commented:
sql has some default values that are taken, (like database) so if i name my database 'database' I'll have synthax errors. 
I made Titles not null cause apparently you can't have a primary with any null values.

Squarific commented:
Your text explains WHAT but not WHY.

Squarific commented on create_db.sql:
```diff
@@ -0,0 +1,9 @@
+CREATE DATABASE IF NOT EXISTS `database_note`;
```
Why an underscore?

cheungkinwong replied:
force of habbit from js where variables can't have spaces.

cheungkinwong replied:
force of habit from js where variables can't have spaces

Squarific commented on create_db.sql:
```diff
@@ -0,0 +1,9 @@
+CREATE DATABASE IF NOT EXISTS `database_note`;
+
+CREATE TABLE IF NOT EXISTS `database_note`.`notes` ( 
+     `Dates` TIMESTAMP on update CURRENT_TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP ,
```
Why do you start your fields with a capital letter?
What is Dates? Why is it plural? Why is it called Dates? Would 'lastupdated' or something not be clearer?

cheungkinwong replied:
I wanted to use date first but that gave an error so i lazily added an s instead.
but yes last_updated would be clearer.

cheungkinwong replied:
Force of habit.
I wanted to use date first but got an error, so i lazily added s.
but indeed last updated would be more clear.

Squarific replied:
In SQL you can use reserved words if you put them between backticks normally.

Squarific commented on create_db.sql:
```diff
@@ -0,0 +1,9 @@
+     `Dates` TIMESTAMP on update CURRENT_TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP ,
+     `Titles` VARCHAR(40) NOT NULL ,
+     `Author` VARCHAR(40) NOT NULL ,
+     `Notes` TEXT NULL DEFAULT NULL , 
```
Why is this plural?
Will a limit be enforced on the PHP side?

cheungkinwong replied:
No i don't think I'll enforce a limit because i want the user to be able to make multiple notes of the same title.

Squarific replied:
This is about the length of the field though. Are you going to put multiple notes into one field?

Squarific commented on create_db.sql:
```diff
@@ -0,0 +1,9 @@
+CREATE TABLE IF NOT EXISTS `database_note`.`notes` ( 
+     `Dates` TIMESTAMP on update CURRENT_TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP ,
+     `Titles` VARCHAR(40) NOT NULL ,
+     `Author` VARCHAR(40) NOT NULL ,
```
Why is this singular? Why the limit of 40 chars?

cheungkinwong replied:
to prevent the title and author from being too long.

Squarific replied:
And why 40 or 20 specifically? Why is 42 chars too long?

Squarific commented on create_db.sql:
```diff
@@ -0,0 +1,9 @@
+     `Titles` VARCHAR(40) NOT NULL ,
+     `Author` VARCHAR(40) NOT NULL ,
+     `Notes` TEXT NULL DEFAULT NULL , 
+     PRIMARY KEY (`Titles`)
```
Why is titles the primary key? Might there be a need in the future for duplicate titles?

cheungkinwong replied:
I think the user should be allowed to make multiple notes with the same title.
for example 'groceries' should be allowed for different days.
I'd probably need to make the primary constraint include, author and dates.
I will research that now.

Squarific replied:
If you want multiple notes with the same title you cannot make it a primary key though.
I also think you can only have a primary key on one field.

cheungkinwong replied:
according to w3school, the value of a primary key can be made up of multiple columns.

cheungkinwong committed:
changed primary key

Squarific commented on create_db.sql:
```diff
@@ -1,9 +1,9 @@
-CREATE DATABASE IF NOT EXISTS `database_note`;
+CREATE DATABASE IF NOT EXISTS `database note`;
```
Why is there a space here?

Squarific commented on create_db.sql:
```diff
@@ -1,9 +1,9 @@
-     `Author` VARCHAR(40) NOT NULL ,
+CREATE TABLE IF NOT EXISTS `database note`.`notes` ( 
+     `Last updated` TIMESTAMP on update CURRENT_TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP ,
+     `Title` VARCHAR(20) NOT NULL ,
```
Why is the title limited to 20 characters?

Squarific reviewed (commented):
-

cheungkinwong committed:
renamed titles according to conventions

cheungkinwong commented:
Changed the code according to feedback and conventions.
I changed all to lower case and removed white spaces to avoid having to use quotation marks.
I used underscore instead to keep the readability.
Also changed the names to avoid plurals to make things more consistent and unambiguous.
The limit in characters are to prevent performance issues. 32 is a (2^5) which is convention according to some comment in stackoverflow.
